src/hardware/pcspeaker_impulse.h

```diff
@@ -0,0 +1,110 @@
+/*
+ *  SPDX-License-Identifier: GPL-2.0-or-later
+ *
+ *  Copyright (C) 2022-2022  The DOSBox Staging Team
+ *
+ *  This program is free software; you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation; either version 2 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License along
+ *  with this program; if not, write to the Free Software Foundation, Inc.,
+ *  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ */
+
+#include "pcspeaker.h"
+
+#include <array>
+#include <deque>
+
+#include "inout.h"
+#include "setup.h"
+#include "soft_limiter.h"
+#include "support.h"
+#include "pic.h"
+
+class PcSpeakerImpulse final : public PcSpeaker {
+public:
+	PcSpeakerImpulse();
+	~PcSpeakerImpulse() final;
+
+	void SetFilterState(const FilterState filter_state) final;
+	void SetCounter(const int cntr, const PitMode pit_mode) final;
+	void SetPITControl(const PitMode pit_mode) final;
+	void SetType(const PpiPortB &port_b_state) final;
+
+private:
+	void AddImpulse(float index, const int16_t amplitude);
+	void AddPITOutput(const float index);
+	void ChannelCallback(uint16_t requested_frames);
+	void ForwardPIT(const float new_index);
+	float CalcImpulse(const double t);
+	void InitializeImpulseLUT();
+
+	// Constants
+	static constexpr char device_name[] = "PCSPEAKER";
+	static constexpr char model_name[]  = "impulse";
+
+	static constexpr int16_t positive_amplitude = 20000;
+	static constexpr int16_t negative_amplitude = -positive_amplitude;
+	static constexpr int16_t neutral_amplitude  = 0;
+
+	static constexpr float ms_per_pit_tick = 1000.0f / PIT_TICK_RATE;
+
+	// Mixer channel constants
+	static constexpr uint16_t sample_rate        = 32000u;
+	static constexpr uint16_t sample_rate_per_ms = sample_rate / 1000u;
+
+	static constexpr auto lowpass_cutoff_freq = 4300;
+	static constexpr auto minimum_counter     = 2 * PIT_TICK_RATE / sample_rate;
+
+	// must be greater than 0.0f
+	static constexpr float cutoff_margin = 0.2f;
+
+	// Should be selected based on sampling rate
+	static constexpr float high_pass_amount   = 0.999f;
+	static constexpr uint16_t filter_quality      = 100u;
+	static constexpr uint16_t oversampling_factor = 32u;
+	static constexpr uint16_t filter_width        = filter_quality * oversampling_factor;
```

Comment:
Could you prefix these with "impulse" or "sinc"? It was really apparent first what kind of filtering they're referring to at first sight.